src/frontend/src/binder/set_expr.rs

```diff
@@ -148,6 +148,20 @@ impl Binder {
                             .into());
                         }
 
+                        // Handle type alignment for values union values
+                        // E.g. Select 1 UNION ALL Select NULL
+                        match (&mut left, &mut right) {
+                            (BoundSetExpr::Select(l_select), BoundSetExpr::Select(r_select)) => {
+                                for (i, (l, r)) in l_select.select_items.iter_mut().zip_eq_fast(r_select.select_items.iter_mut()).enumerate() {
+                                    let expr = align_types(vec![l, r].into_iter())?;
+                                    l_select.schema.fields[i].data_type = expr.clone();
+                                    r_select.schema.fields[i].data_type = expr;
+                                }
+                            },
+                            // TODO: handle other cases
+                            _ => {},
```

Comment:
We may not mark the issue as resolved (or may need to open a new one) for this TODO:
```
values (1) union values (1.2);
```

For non-select, the type of `null` is fixed to `varchar`, but we still need to align the types for others. And in some cases it is more tricky:
```
select 1 union all select 2 union all select 1.5;
```